### ISSUE TITLE
Tool-Jev follow-up: close gaps to a stronger Jev-style decision model

### ISSUE THREAD

**Issue by OriNachum** [agentculture/nvsh#53]
## Context

PR #52 produced two useful Qwen3.5-0.8B Tool-Jev variants:

- **Track A** — generative structured tool caller.
- **Track B** — one-position candidate scorer: list candidate actions under one-token labels, read the next-token scores, normalize them over the offered candidates, then let deterministic nvsh code ground/validate arguments.

Track B is meaningfully **Jev-like** in its runtime shape, but there are still important gaps between what we have and the stronger "System One / Jev-style decision model" abstraction. This issue records those gaps and points to the code and external references to investigate.

This is a **follow-up to #46 / PR #52**, not a claim that PR #52 failed. The PR established the baseline we need to test these questions properly.

## Where to look in PR #52

Core files:

- `scripts/lfm-finetune/scorer.py`
  - Track B prompt, fixed label mapping, next-token scoring, normalization, deterministic argument grounding.
- `scripts/lfm-finetune/train_scorer.py`
  - label-restricted cross-entropy at the single decision position.
- `scripts/lfm-finetune/metrics.py`
  - ECE/Brier and abstention metrics.
- `scripts/lfm-finetune/track_a_calibration.py`
  - Track A's offline teacher-forced candidate distribution.
- `docs/benchmarks/2026-09-24-qwen-tool-jev-comparison.md`
  - final comparison, calibration results, missing-candidate slice and edge results.
- Issue #46
  - original Tool-Jev experiment definition.

PR: https://github.com/agentculture/nvsh/pull/52

## Gap 1 — semantic `escalate` vs uncertainty-driven abstention

Today Track B includes `escalate` as just another candidate label.

Conceptually:

```text
P(service_status)
P(container_list)
P(power_set)
P(explain)
P(escalate)
```

That teaches **"escalate is the correct semantic class for this request"**.

It does **not** directly implement:

> none of the offered actions has enough probability / margin to justify acting.

Those are different states and should probably remain different:

- **semantic escalation** — the task genuinely belongs to a stronger agent/human.
- **epistemic abstention** — the scorer is uncertain about the available choices.

PR #52's missing-candidate slice exposes the distinction:

- Track B escalates only **7/32** when the correct operation is removed.
- It usually chooses a plausible remaining operation instead.

That may be partly a data gap, but it is also an architectural question: a closed softmax classifier is expected to put its mass somewhere.

### Experiment

Keep `escalate` as a real semantic candidate, but add a **runtime uncertainty gate** derived from the distribution.

Candidate rules to test:

```python
if p_escalate >= semantic_escalation_threshold:
    escalate()

elif p_top1 < min_probability:
    abstain_uncertain()

elif p_top1 - p_top2 < min_margin:
    abstain_uncertain()

else:
    propose(top1)
```

Potentially also test entropy / normalized entropy.

Measure thresholds independently for read-only vs mutating actions; the cost of a false positive is not equal.

### External reference

TypeSafe describes Jev as returning typed decisions with calibrated probabilities so **application code can set thresholds for autonomous action vs review/escalation**:

- https://typesafe.ai/
- https://typesafe.ai/blog/introducing-system-one-models-and-jev

## Gap 2 — fixed label identity may let us learn a classifier shortcut

Current Track B preserves one stable mapping from operation identity to token:

```python
label = LABEL_ALPHABET[full.index(name)]
```

So, for example, if `service_status` is always label `A`, the model can partially learn:

```text
service-ish request -> token A
```

instead of the more general behavior:

```text
read the candidates supplied in this request
compare them against the state
select the best candidate
```

The current scorer therefore proves that one-position classification works, but does **not yet prove runtime-defined candidate reasoning**.

### Highest-priority experiment

Evaluate the existing scorer with:

1. candidate order shuffled;
2. operation -> letter mapping randomly permuted;
3. subsets of candidates presented;
4. semantically equivalent candidate descriptions with different labels.

Do this **before retraining** to measure how much the model depends on fixed label identity.

Then train a variant where candidate -> letter assignment is randomized **per training example**.

Example:

```text
sample 1:
A = power_set
B = service_status

sample 2:
A = service_status
B = container_list

sample 3:
A = container_list
B = power_set
```

If performance survives this, we have much stronger evidence that the model learned the Jev-like operation:

> evaluate this state against these supplied alternatives.

### External reference

OpenJev is particularly relevant because its public implementation uses a similar first-output-position technique, but:

- labels/options live in the request;
- it reads one score per option at the first output position;
- it explicitly trains/tests option-order stability;
- it reports only ~2.3% answer changes when option order is shuffled.

See:

- https://huggingface.co/openjev/openjev

This is probably the most directly comparable implementation to Track B.

## Gap 3 — normalized scores are not automatically calibrated correctness probabilities

Track B computes a valid normalized distribution over offered labels, but that alone does not mean:

> `p=0.8` means this decision is correct about 80% of the time.

PR #52 correctly measures calibration instead of assuming it:

- `scorer-b1` test ECE: **0.132**
- Brier: **0.268**
- the ECE success bar was <= 0.10, so it failed.
- the shipped quantized scorer currently has no directly measured complete calibration distribution because the serving path does not return every label score.

So the current wording should remain conservative:

> candidate scores normalized into a decision distribution, with calibration measured separately.

### Work to do

- Find a serving/readout path for quantized builds that returns **all offered label logits**, not top-k.
- Calibrate the **actual deployed artifact**, not only bf16.
- Evaluate temperature scaling / vector scaling / other simple post-hoc calibration first.
- Calibrate separately by:
  - read-only vs mutating;
  - candidate count;
  - missing-candidate / OOD slice;
  - confidence bucket.
- Report reliability diagrams in addition to ECE/Brier.

### External references

TypeSafe says Jev is explicitly optimized for calibrated decisions using **Reinforcement Learning for Calibrated Decisions (RLCD)** and describes the output as epistemically honest probabilities:

- https://typesafe.ai/blog/introducing-system-one-models-and-jev
- https://typesafe.ai/

OpenJev is also useful here because its public readout uses fixed calibration settings after reading first-position option scores:

- https://huggingface.co/openjev/openjev

We do **not** need to copy RLCD to close this gap; first determine how far standard calibration gets us.

## Gap 4 — Track A is Jev-inspired, not a native Jev-style decision model

Track A is still autoregressive generation:

```text
state -> Qwen -> structured tool call
```

`track_a_calibration.py` reconstructs a candidate distribution offline by teacher-forcing each candidate continuation. That is useful evaluation, but the distribution is not Track A's native runtime interface.

Keep Track A — it is useful and achieved the best raw tool-call accuracy — but distinguish it in documentation:

- **Track A:** specialized generative tool router.
- **Track B:** Jev-style candidate decision/scoring model.

This avoids implying the two architectures are equally "Jev-like".

Reference:

- `scripts/lfm-finetune/track_a_calibration.py`
- TypeSafe's own framing: "unstructured state in, typed probabilistic decisions out", giving up string generation:
  https://typesafe.ai/blog/introducing-system-one-models-and-jev

## Gap 5 — Tool-Jev currently implements one narrow Choice primitive

Track B is effectively:

```text
Choice<operation | explain | escalate>
```

That is enough for nvsh's immediate routing problem, and we should **not** add abstractions just to imitate another API.

But if this becomes a reusable local decision model, keep the architecture open to the three useful decision shapes:

- **Choice** — distribution over discrete alternatives.
- **Noul / binary judgment** — probability of yes/true.
- **Score** — distribution over ordered levels / derived scalar.

Potential nvsh examples:

```text
choice:
  Which recovery operation fits?

binary/noul:
  Is retrying this command safe?

score:
  How risky is executing this proposed mutation?
```

This could allow multiple narrow decisions to compose in deterministic code instead of turning Tool-Jev into another miniature agent.

References:

- TypeSafe System One / Jev:
  https://typesafe.ai/blog/introducing-system-one-models-and-jev
- Their workflow decomposition examples:
  https://evals.typesafe.ai/
- OpenJev exposes compatible `choice`, `noul`, and `score` shapes:
  https://huggingface.co/openjev/openjev

## Gap 6 — runtime-defined candidates should become a first-class capability

Track B already accepts an `offered` subset in the scorer, which is a good seam, but training is still based on the fixed nvsh operation set and fixed token identity.

Longer-term target:

```text
state
+ arbitrary bounded candidate descriptions supplied at runtime
-> probability distribution
```

rather than:

```text
state
-> classifier over operations known during fine-tuning
```

This does **not** mean Tool-Jev must become a general model. A specialized nvsh decision model can still be trained only on shell/operations while learning to compare runtime-supplied candidate descriptions.

The label-randomization experiment above is the first test of whether this is already latent in the current model.

## Proposed order

1. **Permutation test the existing `scorer-b1`** — no retraining.
2. Add uncertainty-abstention thresholds and re-run missing-candidate + wrong-mutating metrics.
3. Train one scorer with **per-example randomized candidate labels/order**.
4. Add a complete-logit readout for the deployed Q4 model and calibrate the real artifact.
5. Only then decide whether we need:
   - extra missing-candidate data;
   - explicit OOD/no-valid-option objective;
   - alternative loss / calibration training;
   - generic Choice/Noul/Score API.

## Success criteria for the next iteration

A stronger Tool-Jev should demonstrate:

- [ ] high accuracy when candidate labels/order are permuted;
- [ ] strong behavior when the correct candidate is missing;
- [ ] semantic `escalate` distinguished from uncertainty abstention;
- [ ] 0 wrong mutating actions at the chosen action thresholds;
- [ ] calibrated probabilities measured on the actual deployed/quantized artifact;
- [ ] candidate descriptions supplied at runtime materially influence the decision;
- [ ] no free-form generation on the Track B path;
- [ ] deterministic code remains responsible for grounding, validation and execution.

## Summary

PR #52 established the important part: **Qwen3.5-0.8B can be turned into a very small, fast one-position tool decision model.**

The next question is more interesting than improving ordinary classification accuracy:

> Can we make it genuinely evaluate runtime-supplied alternatives and expose uncertainty that nvsh can safely act on?

That is the gap between "tiny tool classifier" and a stronger local Jev-style decision primitive.

**Comment by OriNachum** [comment]
## Design note — the interesting result may be deeper than “tiny tool classifier”

One insight worth preserving from the review:

> **PR #52 may already contain the seed of a genuinely Jev-like local decision primitive — but we have not yet proved whether the model learned “evaluate supplied alternatives” or merely “classify requests into familiar fixed labels.”**

That distinction is important because both systems can look identical on the current benchmark.

Today Track B does:

```text
state + candidate descriptions
    ↓
Qwen3.5-0.8B
    ↓
one-position logits over candidate labels
    ↓
distribution + deterministic grounding
```

At first glance this is just a classifier. But if the scorer still works when candidate identities are **remapped to arbitrary labels at runtime**, then the model has learned something substantially more general:

> compare the current state against the alternatives described in this request, then score those alternatives.

That is much closer to the interesting Jev/System-One property than simply predicting a fixed operation ID.

### Why the permutation experiment matters so much

Suppose training always contains:

```text
A = service_status
B = container_list
C = power_set
```

A successful model may have learned either:

```text
service-ish request -> A
```

or:

```text
read what A means
read what B means
read what C means
compare each with the request
choose A
```

Ordinary accuracy cannot distinguish those hypotheses.

A permutation test can.

For the same request, evaluate:

```text
A = service_status
B = container_list
C = power_set
```

then:

```text
A = power_set
B = service_status
C = container_list
```

then a reduced or reordered candidate set.

If the *semantic operation* remains stable while the emitted letter changes correctly, that is evidence the model is actually evaluating the supplied candidate descriptions rather than relying on a learned token/class identity.

This makes the permutation test more than another robustness benchmark: it is an **architectural probe into what computation the fine-tune learned**.

### A second important distinction: two kinds of “I should not act”

The current `escalate` candidate mixes easily with a separate concept that the runtime probably needs:

1. **Semantic escalation**
   - “This task genuinely belongs to a stronger agent / human.”
   - This is a legitimate semantic class for the model to score.

2. **Epistemic abstention**
   - “I cannot distinguish the offered actions confidently enough.”
   - This should emerge from the *shape of the distribution* and application policy, not necessarily from another learned class.

For example:

```text
service_status   0.41
container_list   0.38
power_set        0.08
...
```

may be very different from:

```text
escalate         0.87
service_status   0.05
...
```

The first says “ambiguous among available actions”; the second says “escalation itself is the best semantic action.”

Keeping those concepts separate lets deterministic nvsh policy decide when uncertainty is acceptable, especially for mutating actions.

### The probability nuance is also part of the result

The Track B values are currently best understood as:

> **normalized evidence over the offered alternatives**

not automatically:

> **probability that the chosen action is objectively correct.**

The latter requires calibration evidence.

PR #52 already did the right thing by measuring ECE/Brier instead of assuming softmax = truth probability. That measurement discipline is part of what makes this experiment valuable.

If the next iteration combines:

- runtime-defined candidate semantics;
- arbitrary label permutation;
- full candidate logits;
- calibration on the deployed quantized artifact;
- an external uncertainty gate;

then Tool-Jev becomes something more interesting than a small router:

```text
fuzzy local judgment
    ↓
bounded probabilistic decision
    ↓
deterministic policy / grounding / execution
```

That separation is arguably the strongest design idea to carry forward from #52.

### What would count as the strongest result

A particularly compelling result would be:

> A 0.8B local model, trained for a narrow operational domain, can accept a changing bounded set of action descriptions, assign useful calibrated scores to those actions in one forward decision position, retain the same semantic choice under arbitrary label permutation, and defer execution/uncertainty policy to deterministic code.

That claim is much stronger — and more interesting — than:

> “we fine-tuned Qwen3.5-0.8B to call nvsh tools.”

It is also experimentally falsifiable, which makes the next steps very clear.

**Comment by OriNachum** [comment]
Scoped and specced: `docs/specs/2026-09-24-tool-jev-calibration-cycle-issue-53.md` on branch `spec/tool-jev-calibration-issue-53` (frame `tool-jev-calibration-cycle-issue-53`, challenge pass done). Track B is the subject; the cycle is experiment-only (`scripts/lfm-finetune`), with the permutation bar at OpenJev's 2.3%.

Split out as follow-ups:

- #54 wire the calibrated scorer into the tiers as a Verifier (runtime)
- #55 Choice / Noul / Score primitives (gap 5)
- #56 Track A follow-ups
- #57 `measure.py` failed start-up blocks a clean re-run (P71)
- #58 `serve_for_measure.sh` same-port stop/start race
- #59 corpus v2 promotion into `nvsh/tiers/corpus` and the public held-out
- #60 close out #46 records and per-repo Hub visibility

RLCD is not copied (TypeSafe has not published the method); a calibration-aware training stage of our own runs only if post-hoc scaling plus new data miss ECE <= 0.10.

- nvsh (Claude)
